File: ABC/abc324/c.cpp
#include <iostream>
#include <vector>
#include <set>
#include <string>
#include <deque>
using namespace std;

int main() {
    int N;
    string Tp;
    cin >> N >> Tp;

    vector<string> S(N);
    for (int i = 0; i < N; i++) {
        cin >> S[i];
    }

    set<string> setT;
    int ltp = Tp.length();

    // Tp[:x]とTp[x:]を計算
    vector<string> Tptox(ltp + 1);
    vector<string> Tpfromx(ltp + 1);
    deque<char> top, bottom(Tp.begin(), Tp.end());
    
    for (int i = 0; i <= ltp; i++) {
        Tptox[i] = string(top.begin(), top.end());
        Tpfromx[i] = string(bottom.begin(), bottom.end());
        if (i != ltp) {
            top.push_back(Tp[i]);
            bottom.pop_front();
        }
    }

    // Tpから1文字削除したものをsetTに追加
    for (int i = 0; i < ltp; i++) {
        setT.insert(Tptox[i] + Tpfromx[i + 1]);
    }

    // Tpの1箇所にある英小文字1文字を追加したものをsetTに追加
    for (int i = 0; i <= ltp; i++) {
        for (char j = 'a'; j <= 'z'; j++) {
            setT.insert(Tptox[i] + j + Tpfromx[i]);
        }
    }

    // Tpの1箇所をある英小文字1文字に変更したものをsetTに追加
    for (int i = 0; i < ltp; i++) {
        for (char j = 'a'; j <= 'z'; j++) {
            setT.insert(Tptox[i] + j + Tpfromx[i + 1]);
        }
    }

    vector<int> ans;
    for (int i = 0; i < N; i++) {
        if (setT.find(S[i]) != setT.end()) {
            ans.push_back(i + 1);
        }
    }

    cout << ans.size() << endl;
    for (int i = 0; i < ans.size(); i++) {
        cout << ans[i] << " ";
    }
    cout << endl;

    return 0;
}

File: ABC/abc324/c.py
import sys
def input(): return sys.stdin.readline()[:-1]

N, Tp = map(str, input().split())
N = int(N)

S = []
for i in range(N):
  s = input()
  S.append(s)

setT1 = set()
setT2 = set()
setT3 = set()
ltp = len(Tp)

# Tp[:x]とTp[x:]は最初に計算しておく
Tptox = []
Tpfromx = []
from collections import deque
top = deque()
bottom = deque(Tp)
for i in range(ltp+1):
  Tptox.append(''.join(top))
  Tpfromx.append(''.join(bottom))
  if i!=ltp:
    top.append(Tp[i])
    bottom.popleft()

# Tpから1文字削除したもの全てをsetTに追加
for i in range(ltp):
  setT1.add(''.join([Tptox[i], Tpfromx[i+1]]))

# Tpの1箇所にある英小文字1文字を追加したもの全てをsetTに追加
for i in range(ltp+1):
  for j in range(26):
    setT2.add(''.join([Tptox[i], chr(ord('a')+j), Tpfromx[i]]))

# Tpの1箇所をある英小文字1文字に変更したもの全てをsetTに追加
for i in range(ltp):
  for j in range(26):
    setT3.add(''.join([Tptox[i], chr(ord('a')+j), Tpfromx[i+1]]))

ans = []
for i in range(N):
  if abs(len(S[i])-ltp)>1:
    continue
  if len(S[i])==ltp:
    if S[i] in setT3:
      ans.append(i+1)
  if len(S[i])==ltp+1:
    if S[i] in setT2:
      ans.append(i+1)
  if len(S[i])==ltp-1:
    if S[i] in setT1:
      ans.append(i+1)

print(len(ans))
print(*ans)
